### ISSUE TITLE
[Proposal] Add IClusterClient.GetGrains and IStorageProvider.ReadStates (plural)

### ISSUE THREAD

**Issue by nkosi23** [dotnet/orleans#7660]
**Proposal at a conceptual level**

Currently Orleans grains can only be used to retrieve a reference to a single Grain, however it would be beneficial to be able to retrieve multiple grains to support database queries based on various filters (and not just on retrieving an entity by id). To clarify: this proposal is not about adding filtering and aggregation capabilities to Orleans.

To give an example, we use a NoSQL document database (CouchDB) giving us the option to either return full documents or the ids of the documents matching a query. My idea is to be able to pass these ids to Orleans so that it can retrieve all relevant documents (grain state) in one network roundtrip. 

One may think that creating an extension method would be enough, however it would be inefficient as considering the current API exposed by Orleans one would need to loop through all document ids and retrieve them one by one. The idea would therefore be to have Orleans read states in a single network round-trip internally (thus the need for an IStorageProvider.ReadStates method).

In practice, Orleans would check if the grain matching a particular id is already activated, and if so return the already existing activation. Otherwise the id is added to the list of ids that will be passed to IStorageProvider.ReadStates for bulk activation.

**What would be the benefits**

At present, query scenarios intended to retrieve more than one entity require users to bypass Orleans and query the database directly. The main benefit of this feature would be enabling users to more extensively use Orleans as a dynamic cache reducing the load on the database for dynamic query, especially since it is very common for applications to see their queries always return the same results in practice for a period of time. 

For example, the latest articles of a newspaper do not change much over the course of a day. The catalog of products of an e-commerce store is also fairly static. Similarly, the latest content relevant to a user typically remains the same for a period of time in a wide range of applications.

While this feature will not automatically reduce the number of round-trips to the database, it will substantially reduce memory use on the database, and the payload of responses and therefore increase throughput. Moreover, by adding an additional external cache between the  database and the application to cache specific queries, one will be able to take the number of requests hitting the database down to zero while presenting up-to-date data to the user (however this is less useful for queries being very user-specific as the size of the cache would grow too large). Even without the external cache, this feature would already be extremely useful.

The other benefit would be enabling users to fully use Orleans as the single source of truth for the application even in such query scenarios,  otherwise the data may potentially be stale if an in-memory change has taken place in the grain without being committed to the data store. Even if the result set returned by the database may not always be fully correct in such a scenario if the filtering was made on a database column changed in-memory in the meantime, this is already massively helpful for a vast range of scenarios (including the most common ones).

**Comment by JorgeCandeias** [comment]
Isn't this [Orleans.Indexing](https://github.com/OrleansContrib/Orleans.Indexing)?

**Comment by nkosi23** [comment]
@JorgeCandeias Wow this a very exciting project, wasn't aware of it. My proposal is much simpler however, I only request the ability to fetch mutiple grains with their state in a single network round-trip using their id.

For simplicity I am not requesting the ability to create and use arbitrary indexes in orleans - but I agree this would be great. This indexing project could also make use of the API I describe here if it ever needs to activate grains not currently activated.

**Comment by JorgeCandeias** [comment]
@nkosi23 

This is an interesting challenge. I agree it would be nice to have support for querying distributed datasets in one go, however the way one goes about it matters.

- If we read data straight from storage, won't we fall into the anti-pattern of treating grains/actors like data instead of agents with ownership of their own data? In principle, internal grain state should not be accessible without having to go through the grain, in order to avoid pulling the rug from under the grain.
- And if do treat grains like we do "entities", aren't we also confusing *public grain data* with *internal grain state* with *shared application state*. They're different things, even if they appear to overlap some times.
- On the other hand, if we do go through the grain, aren't we risking having to activate thousands of grains just to get the data, which was the initial problem?
- And if we index or aggregate the data somewhere else, how do we balance it across the cluster to avoid creating hot spots that bottleneck performance?

This a catch-22 in actor systems and it's hard to address. I've seen developers treat actors like entities, as they're still thinking in stateless terms, and you end up with poor designs that do little more than read caching and have lots of round-trips for no good reason. One may as well just query the database and add some more memory there, it achieves the same.

This catch-22 is something I suspect the indexing project was attempting to solve by focusing on indexing some "public grain properties" only (via some user class similar to how state works), and not wholesale touching the rest of the internals, which respects the actor model principle of actors owning their state. It's by far not the easy way to solve the problem, but I think it's the more appropriate way to solve the problem in this context. Shame it's not getting much traction, it's remarkable tech.

> My idea is to be able to pass these ids to Orleans so that it can retrieve all relevant documents (grain state) in one network roundtrip.

As commented above, isn't this a case of confusing internal grain state (some dotnet objects) with application state (the documents). Internal grain state is meant to support the grain lifecycle and responding to methods called on it from memory. Shared application state is the whatever persisted data your entire application keeps. They're not the same thing in concept, and often not the same in practice either, if one is following DDD design principles by keeping distinct business and data layer models separate from each other.

Case in point, what do you do when:

- the grain has multiple shards of data;
- or you need data from different types of grains;
- or that persisted data doesn't match the interface supplied data, because the grain massages it somehow;
- or what if the grain loads data from somewhere else entirely?

Doesn't this sound like a can of worms?

> Moreover, by adding an additional external cache between the database and the application to cache specific queries, one will be able to take the number of requests hitting the database down to zero while presenting up-to-date data to the user (however this is less useful for queries being very user-specific as the size of the cache would grow too large)

This is what the reactive caching pattern is for, by the way. You can combine it with streaming to consume incoming data from multiple other grains and aggregate/group/organize that into projection grains which are ready for querying. The pattern even works with differential collection updates, here's an [advanced example](https://github.com/JorgeCandeias/Trader/blob/master/Trader.Trading/Providers/Trades/TradeProviderReplicaGrain.cs) of collection replication. Granted this is application work, not an Orleans built-in feature and you may still end up with hotspots if the projection grains aren't sharded properly. I've done plenty of work with reactive caching and projections, I can probably help you out in more detail if you want to follow this route.

Now I'm not claiming to have a solution to the catch-22 either, just babbling on my thoughts here. I would have liked the indexing project to be an active component of Orleans, as it would solve both your problem and make many others easier but alas, it's not the case right now.

**Comment by nkosi23** [comment]
@JorgeCandeias Thank you a lot for taking the time to explain this to me! This is humbling and reminds me that I still have a lot to learn when it comes to the virtual actor pattern. In particular, I didn't realize that treating grains like data instead of agents is an anti-pattern (i didn't fully see them as data but I assumed it was fine to see them that way on occasions. I now realize that this is actually dangerous). However, while I see the nuance and can sense what I need to watch out with regards to this point, when you add that by treating actors like entities one is confusing grain state with application state I am sort of overwhelmed and may need some help to understand :sweat_smile:

In my mind, an actor = a DDD entity, but lately I have also finally understood the remark you gave me in some other ticket a while ago when you told me that grains are also a great place to put business logic. What I have understood is that grains are more than entities in that it is also very relevant to put all the business logic relating to an entity in a grain (use cases, workflows, application services) while an entity is focused on managing its own internal state.

Still, I am currently considering that Grain State = my Entity state. To me a grain is some sort of higher order entity being also able to orchestrate business logic relating to an entity type. It may have additional methods to better support the life of the entity (from a business perspective, not from a runtime/technical perspective), but the state of the grain is the one a particular entity. This is my understanding.  

But if I understand correctly you are telling me that in good designs, it is very common to see that the state of the Actor isn't the state of the entity, but actually has different concerns. So I am wondering what are the concerns of an actor then? What other concerns could it have than managing the entity it represents (is its role even to represent an entity then?). 

I understand that for simple systems actor state may be amalgamated with entity state safely, but I have sense that this is similar to saying that for simple CRUD web applications, one may "safely" use the very object representing the database model as a view model. As a DDD practitioner I am aware that this stops being a good idea very rapidly so I feel a need to better understand the distinction you make between actor state and entity state / application state if this analogy is true. :smile: 

I believe the reasons I've been bitten by this is that I usually cheat a little with DDD by using entities as my persistence objects to save on code duplication. I use [the Onion Architecture](https://jeffreypalermo.com/2008/07/the-onion-architecture-part-1/) and JSON.NET is the only dependency I tolerate in the Domain Model project. With the type of projects I worked on, I haven't faced yet a situation where I wished I had separated the domain model from persistence (but this may due to the fact that I always have the freedom to use a NoSQL document database on my projects and as such there is no impedance mismatch between the domain model and the entity, so I considered it would be more rational to make this compromise). Now the problem is that I have approached Orleans with this background and mindset and it looks like this will bite me :sweat: 

Regarding the feature I have proposed, you are making a good point that thousands of grains may have to be activated at once. While in my case i often need to load the full document anyway (one of the reasons being that my backends most of the time expose a web APIs for rich clients that may need to present the data in multiple ways), activating a grain may be way more expensive than fetching a document, and I can also see situations where there may a need to optimize a hot path for performance reason. And I agree that this is can of worms. 

Moreover, in the use case I have presented, I forgot that a grain isn't even supposed/required to load its data from a single data store, it may actually have many. This assumption was probably due to my "actor = entity" mindset. This alone is already enough to make my proposal much less useful (without even adding the other worms coming out of this can).

I now understand that the indexing project is a more sensible way to approach this kind of concern, the first reason for this being that the Indexing project builds on a better understanding of what an actor is meant to be (and what is isn't meant to be). Anyway, you are a very insightful person and have a phenomenal ability to spot where my actual gap of knowledge is, no matter which starting question is on the surface. Thank you so much for your insights. I will definitely look into the pattern and example you have highlighted as it looks useful for a wide range of situations.

**Comment by JorgeCandeias** [comment]
> In my mind, an actor = a DDD entity

TLDR; Change this to "actor = DDD Aggregate Root" and everything will make sense.

Long story:
- Entity: This is some object that has an *identity* and has some business meaning, e.g. an individual user, an individual post, etc.
- Value Object: This is some object that is uniquely identified by their properties, e.g. the full address of some house.
- Aggregate: This is some group of entities and/or value objects that we somehow relate to each other in a business sense.
- Aggregate Root: Some other object that owns the aggregate, defines an interaction boundary around it, and acts as an entry point to makes changes on the aggregate. It ensures the many objects the aggregate comprises remain consistent with each other.

Take a look at the [chat sample](https://github.com/dotnet/orleans/blob/main/samples/ChatRoom/ChatRoom.Service/ChannelGrain.cs). The Channel grain manages two sets of objects:
- a list of members online.
- a cache of the latest messages in the chat room.

The individual members are entities. The individual messages are also also entities. But the chat room grain is not an entity. It is an aggregate root, which is managing a group of different entities and keeping them consistent within the boundary of a chat room.

Does this make sense?

> But if I understand correctly you are telling me that in good designs, it is very common to see that the state of the Actor isn't the state of the entity, but actually has different concerns. So I am wondering what are the concerns of an actor then?

When you are writing notes on a piece of paper, you are the agent of writing, the paper is the medium of writing, and the notes are the content of writing.

Grains are the agent, not the medium and not the content. Treating them as the latter misses out on the capabilities of the former.
 
> I understand that for simple systems actor state may be amalgamated with entity state safely, but I have sense that this is similar to saying that for simple CRUD web applications, one may "safely" use the very object representing the database model as a view model.

This stops being safe the moment you discover some internal private data made its way to the public swagger API because someone added some new property to an internal model. I wish I'd gotten a bitcoin every time I saw this happen, back when bitcoin was under a dollar. 💵 💵 💵 

> Now the problem is that I have approached Orleans with this background and mindset and it looks like this will bite me 😓

It may look verbose and overkill but if we're doing something important, it's best to split the three model layers upfront, even if they may look redundant copy-paste at the start of a project. We can use something like automapper to reduce the pain. Doing so can outright avert mass technical debt when the solution grows more complex.

> Regarding the feature I have proposed, you are making a good point that thousands of grains may have to be activated at once. While in my case i often need to load the full document anyway (one of the reasons being that my backends most of the time expose a web APIs for rich clients that may need to present the data in multiple ways), activating a grain may be way more expensive than fetching a document, and I can also see situations where there may a need to optimize a hot path for performance reason. And I agree that this is can of worms.

This is why the way the Orleans team go about this is important. They need to solve this for the common denominator, not for specific use cases. Good, because that's our job and it keeps us fed and clothed. 😄
